### PR TITLE
Fix #1151: DataExporter calls load method twice on lazy DataTables; "Plan B"

### LINKED ISSUE #763: Exporting from Data Table with lazy data model generates incorrect file if table is manipulated while exporting
The `org.primefaces.component.export.Exporter.exportAll(FacesContext, DataTable, Object)` method manipulates the `DataTable` object it receives, establishing the first row as 0, the page size as the total number of records, then it loads the "page" (the whole list) and it exports it, row by row. Finally, it restores the original values of the different variables and loads again the original page the user was seeing.

``` java
if(lazy) {
        if(rowCount > 0)
             {
                table.setFirst(0);
                table.setRows(rowCount);
                table.clearLazyCache();
                table.loadLazyData();
             }

            for(int rowIndex = 0; rowIndex < rowCount; rowIndex++) {
                exportRow(table, document, rowIndex);
            }

            //restore
            table.setFirst(first);
            table.setRowIndex(-1);
            table.clearLazyCache();
            table.loadLazyData();
        } 
```

The problem is that sometimes a lazy export can take long time to execute (why would we use lazy tables if not?) and, while the file is being generated, the user could change the filter or change the page, which would change the same `DataTable` object the export is using, thus generating an incorrect file.

Right now, my workaround is to create an identical hidden DataTable and linking the exporter to that one (since the user cannot manipulate it).

I'm guessing a good solution would be to clone the `LazyDataModel` object and use this clone only for that particular export. I guess the main challenge should be to evaluate the EL expressions over items of the clone instead of over the original one (maybe cloning also the `DataTable` object?).

I'm using PF 5.1 but checking [https://github.com/primefaces/primefaces/blob/master/src/main/java/org/primefaces/component/export/Exporter.java](https://github.com/primefaces/primefaces/blob/master/src/main/java/org/primefaces/component/export/Exporter.java), the bug seems to be also present on the latest version.

Thank you very much.

### LINKED ISSUE #1151: DataExporter calls load method twice on lazy DataTables
I'm using a DataTable with `lazy="true"` to paginate a big datasource. It works perfect, but, if you include a `<p:dataExporter>` for that table, the query is executed twice while exporting the data (without pagination).

You can test it with the lastest version of the Showcase, just include a dataExporter on \src\main\webapp\ui\data\datatable\lazy.xhtml

``` xml
<p:dataTable var="car" value="#{dtLazyView.lazyModel}" paginator="true" rows="10" id="carTable" lazy="true">
    <p:column headerText="Id" sortBy="#{car.id}" filterBy="#{car.id}">
        <h:outputText value="#{car.id}" />
    </p:column>
    <p:column headerText="Brand" sortBy="#{car.brand}" filterBy="#{car.brand}">
        <h:outputText value="#{car.brand}" />
    </p:column>
</p:dataTable>

<h:commandLink value="Export">
    <p:dataExporter type="xls" target="carTable" fileName="test" />
</h:commandLink>
```

In this case, `LazyCarDataModel.load` is called twice (iterating 200 cars each time).

Related to issue #763 
## 

Primefaces 5.3

### PR DESCRIPTION
**Pull request by christophs78**
Showcase-PR: primefaces/showcase-facelift#111

This does not fix #763. This would require some further work. (IMO we can´t just clone LazyDataModel because we don´t know how it was instantiated. The specific implementation of LazyDataModel may contain custom member-variables, ....)

### CODE REVIEW

File: src/main/java/org/primefaces/component/datatable/export/DataTableExporter.java [hunk 1]
```diff
@@ -228,7 +239,13 @@ protected void exportAll(FacesContext context, DataTable table, Object document)
             table.setRows(rows);
             table.setRowIndex(-1);
             table.clearLazyCache();
-            table.loadLazyData();
+
+            if (table.getValue() instanceof LazyDataModel) {
+                lazyDataModel.setWrappedData(wrappedData);
+                lazyDataModel.setPageSize(pageSize);
```
Comment (christophs78) [inline src/main/java/org/primefaces/component/datatable/export/DataTableExporter.java]: Not sure whether  it´s necessary to reset PageSize, RowIndex and RowCount.

Comment (christophs78) [inline src/main/java/org/primefaces/component/datatable/export/DataTableExporter.java]: Not sure whether it´s necessary to reset PageSize, RowIndex and RowCount.

**Comment by Rapster** [comment]
Here is what I did:
```java
        if (lazy) {
            LazyDataModel model = (LazyDataModel) table.getValue();
            List<Object> wrappedData = model.getWrappedData();
            if (rowCount > 0) {
                table.setFirst(0);
                table.setRows(rowCount);
                table.clearLazyCache();
                table.loadLazyData();
            }

            for (int rowIndex = 0; rowIndex < rowCount; rowIndex++) {
                exportRow(table, document, rowIndex);
            }

            //restore
            table.setFirst(first);
            table.setRows(rows);
            table.setRowIndex(-1);
            table.clearLazyCache();
            model.setPageSize(rows);
            model.setWrappedData(wrappedData);
        }
```

No need to check if it's a `LazyDataModel`, because if lazy = true then value must be `LazyDataModel`. That's why I think `DataTable#lazy` is useless and redundant.

Anyway, looks like it's "working". Whatever `DataTable#loadLazyData` is setting, we should do it so I think wr should re-set pagesize, first, etc. 

If it's looks good to you, we should prefer this solution as it's less boilerplate code

**Comment by christophs78** [comment]
It looks good to me!

I tried to replicate what other Code-Fragments do to have consistent code. (eg `DataTable#loadLazyData`, `DataTable#clearLazyCache`, `DataTable#loadLazyScrollData`, ...)
IMO this methods don´t need the instanceof-check too.

I also think DataTable#lazy is useless and redundant. Should we open a issue for 9.0 to deprecate/remove the lazy-attribute (from UIData)?

**Comment by Rapster** [comment]
I'll create a ticket regarding DataTable, this component needs some love 😄 It'll contain few ideas that came on top of my head to clarify and simplify some features. And one of them is of course, removing lazy attribute, but I remember Thomas mention something about not working completely, maybe we'll know what by, looking in the SVN history.

**Comment by christophs78** [comment]
Maybe we should test features like selection or inline-editing against this modification. To be sure we don´t cause a regression a few (AJAX-)requests later because we modified LazyDataModel and only reset wrappedData...

Usecase:
1. User exports table-data
2. a) User selects a row from the current page (https://www.primefaces.org/showcase/ui/data/datatable/selection.xhtml)
b) User switches to a row from the current page into inline-edit-mode (https://www.primefaces.org/showcase/ui/data/datatable/edit.xhtml)

Maybe LazyDataModel need´s some love too. ;-)

**Comment by Rapster** [comment]
Regarding #763 I think blockUI should be used (easiest solution IMO)

**Comment by christophs78** [comment]
`DataTableExporter#exportAll`:

LazyDataModel before `DataTable#loadLazyData`
grafik

LazyDataModel after `DataTable#loadLazyData`
grafik

LazyDataModel after `foreach { DataTable#exportRow }`:
grafik

**Review by melloware — APPROVED** [review]

**Comment by hades646** [comment]
> (IMO we can´t just clone LazyDataModel because we don´t know how it was instantiated. The specific implementation of LazyDataModel may contain custom member-variables, ....)

You could call an abstract method that must be implemented in the custom class. You clone your attributes and the rest must be cloned by the abstract method.

**Pull request merged**